packages/concerto-core/lib/introspect/mapdeclaration.js

```diff
@@ -0,0 +1,182 @@
+     * @throws {IllegalModelException}
+     */
+    constructor(modelFile, ast) {
+        // TODO remove on full release.
+        if(process.env.ENABLE_MAP_TYPE !== 'true') {
+            throw new Error('MapType feature is not enabled. Please set the environment variable "ENABLE_MAP_TYPE=true" to access this functionality.');
+        }
+
+        super(modelFile, ast);
+        this.modelFile = modelFile;
+        this.process();
+    }
+
+    /**
+     * Process the AST and build the model
+     *
+     * @throws {IllegalModelException}
+     * @private
+     */
+    process() {
+        super.process();
+
+        if (this.ast.properties.length !== 2) {
+            throw new IllegalModelException(`MapDeclaration must contain exactly two properties -  MapKeyType & MapyValueType ${this.ast.name}`, this.modelFile, this.ast.location);
+        }
+
+        const key = this.ast.properties.find(property => property.$class === 'concerto.metamodel@1.0.0.MapKeyType');
+        const value = this.ast.properties.find(property => property.$class === 'concerto.metamodel@1.0.0.AggregateValueType' || property.$class === 'concerto.metamodel@1.0.0.AggregateRelationshipValueType');
+
+        if (!key) {
+            throw new IllegalModelException(`MapDeclaration must contain MapKeyType  ${this.ast.name}`, this.modelFile, this.ast.location);
+        }
+
+        if (!value) {
+            throw new IllegalModelException(`MapDeclaration must contain AggregateValueType  ${this.ast.name}`, this.modelFile, this.ast.location);
+        }
+
+        this.name = this.ast.name;
+        this.key = new MapKeyType(this, key, this.modelFile);
+        this.value = new MapValueType(this, value, this.modelFile);
+        this.fqn = ModelUtil.getFullyQualifiedName(this.modelFile.getNamespace(), this.ast.name);
+    }
+
+    /**
+     * Semantic validation of the structure of this class.
+     *
+     * @throws {IllegalModelException}
+     * @protected
+     */
+    validate() {
+        super.validate();
+        this.key.validate();
+        this.value.validate();
+    }
+
+    /**
+     * Returns the fully qualified name of this class.
+     * The name will include the namespace if present.
+     *
+     * @return {string} the fully-qualified name of this class
+     */
+    getFullyQualifiedName() {
+        return this.fqn;
+    }
+
+    /**
+     * Returns the ModelFile that defines this class.
+     *
+     * @public
+     * @return {ModelFile} the owning ModelFile
+     */
+    getModelFile() {
+        return this.modelFile;
+    }
+
+    /**
+     * Returns the short name of a class. This name does not include the
+     * namespace from the owning ModelFile.
+     *
+     * @return {string} the short name of this class
+     */
+    getName() {
+        return this.name;
+    }
+
+    /**
+     * Returns the type of the Map key property.
+     *
+     * @return {string} the short name of this class
+     */
+    getKey() {
```

Comment:
getKeyType would be more consistent with Property

Reply:
We would then be in the territory of defining an api consisting of calls like `clz.getValueType().getType()` which is unclear imo.